### PR TITLE
Improve default bucket count in FTE

### PR DESCRIPTION
**Pull request by dekimir**
## Description

We observed high skew in FTE tasks due to overly low bucket count.

## Release notes

(x) This is not user-visible or is docs only, and no release notes are required.
( ) Release notes are required. Please propose a release note for me.
( ) Release notes are required, with the following suggested text:

```markdown
# Section
* Fix some things. ({issue}`issuenumber`)
```

### CODE REVIEW

**Comment by losipiuk** [comment]
CI: https://github.com/trinodb/trino/issues/18697

File: core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java [hunk 1]
```diff
@@ -235,6 +238,26 @@ private NodePartitionMap systemNodePartitionMap(
         });
     }
 
+    /**
+     * Query plans typically divide data into buckets to help split the work among Trino nodes.  How many buckets?  That is usually dictated by the connector, via
```
Comment (losipiuk) [inline core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java:242]: `usually` - probably not really - but whatever. 
You can rephrase that it is either steered by connector or by engine if connector does not care.

**Review by losipiuk — APPROVED** [review]

File: core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java (hunk 1 above)
Comment (dekimir) [reply core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java]: Dropped "usually".

File: core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java [hunk 2]
```diff
@@ -235,6 +238,26 @@ private NodePartitionMap systemNodePartitionMap(
...
+     *
+     * @return The default bucket count to use when the connector doesn't provide a number.
+     */
+    private int getDefaultBucketCount(Session session, PartitioningHandle partitioningHandle)
+    {
+        // The goal is to have enough buckets to evenly distribute them across tasks created.  The simplest way to do that is to have exactly as many buckets as there are tasks;
+        // then each task gets one bucket.  So let's return the total number of tasks here.  This number, unfortunately, is hard to know precisely because of the variety of ways in
+        // which schedulers create tasks.  We therefore have to estimate.
+        if (getRetryPolicy(session) != RetryPolicy.TASK) {
+            // Pipeline schedulers typically create as many tasks as there are nodes.
```
Comment (sopel39) [inline core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java:253]: this is too low even for pipeline mode as it affects local scaling too

Comment (dekimir) [reply core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java:253]: Can you say more?  What is "local scaling"?  I see that stages schedulers create a task per stage; can this exceed the node count and is it relevant?

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java:253]: I've created follow up PR https://github.com/trinodb/trino/pull/22000

Comment (sopel39) [reply core/trino-main/src/main/java/io/trino/sql/planner/NodePartitioningManager.java:253]: See https://github.com/trinodb/trino/blob/be29ab8ae7b712f9ba66ce169481fc4046292172/core/trino-main/src/main/java/io/trino/operator/exchange/LocalExchange.java#L183 for example. If `bufferCount` is larger than bucket count, then we are not fully utilizing writer concurrency within a task

**Pull request merged**